content/docs/guides/driver-design.md

```diff
@@ -0,0 +1,223 @@
+
+### Constrain memory use at compile time
+
+TinyGo allows for heap allocations and that is a wonderful feature but can also subject embedded Go users to some pain
+if the drivers they use make heavy use of heap allocations. 
+Heap allocations over time can crash your program if memory becomes sufficiently fragmented. This is a huge problem for users who want their programs to run a long time.
+
+
+#### Heap allocations: an introduction
+The most common source of heap allocations in TinyGo drivers are in bus communications. Take for example the following code:
+```go
+type Device struct {
+    spi drivers.SPI
+}
+
+func (d *Device) readRegister(addr uint8) (value uint8, err error) {
+    var data [2]byte
+    err= d.spi.Tx([]byte{addr, 0}, data[:])
+    return data[1], err
+}
+
+
+func (d *Device) writeRegister(addr, newValue uint8) error {
+    var data [2]byte
+    return d.spi.Tx([]byte{addr, newvalue}, data[:])
+}
+```
+
+The code above takes two buffers and passes them to the `drivers.SPI` interface `Tx` method as slices.
+Slices are referential data structures, which is to say, the **data structure contains a pointer**.
+This is a pointer to the start of the data to be sent over the bus.
+
+In go we are not aware of this since we never interact with the low level representation of a slice:
+```go
+type slice struct {
+    data      *type // Points to the first element of the slice:  &s[0]
+    length    int   // This is returned when calling builtin len: len(s)
+    capacity  int   // This is returned when calling builtin cap: cap(s)
+}
+```
+Why is it important to know slices are referential? Well it is because the compile is aware of this and will check
+referential arguments to function and "escape" them if it can't prove their reference (pointer) is not held by the calling function.
+When a reference is escaped it will be forced to be heap allocated by the compiler.
+
+Returning to the `Device` example above we started with: any referential argument to `spi.Tx` will escape since the `Tx` method is unknown at compile time since
+`spi` is an interface. So every time we call `writeRegister` or `readRegister`, at worst two heap allocations will be performed, one being the `data [2]byte` array
+and the second being the inline composite literal slice declaration `[]byte{x, x}`. Usually heap allocators have limitations on size allocated so it is likely that more than 4 bytes are allocated for every call.
+
+#### Heap allocations: Mitigations
+Luckily it can be relatively easy to mitigate heap allocations and eliminate them altogether. To do so with a driver that uses slices for bus communications one can include the array memory within the device struct:
+
+```go
+type Device struct {
+    spi      drivers.SPI
+    bufRead  [2]byte
+    bufWrite [2]byte
+}
+
+func (d *Device) readRegister(addr uint8) (value uint8, err error) {
+    d.bufWrite = [2]byte{addr, 0}
```

Comment:
Does not this create a new slice? I mean, every time. Or is compiler smart enough to replace values in already allocated fixed size slice?

Reply:
I've added a comment to clarify this :)